File: rust-example/src/capabilities/book_stay/decide.rs
use chrono::{DateTime, Utc};
use uuid::Uuid;

use super::context::{
    BookingContext, GuestBookingEligibility, ListingBookingSettings, ListingBookingStatus,
};
use super::request::BookStay;
use super::result::{BookingRejected, ReservationConfirmed};

pub fn decide(
    request: &BookStay,
    context: &BookingContext,
    reservation_id: Uuid,
    now: DateTime<Utc>,
) -> Result<ReservationConfirmed, BookingRejected> {
    if request.stay.check_in >= request.stay.check_out {
        return Err(BookingRejected::InvalidDateRange);
    }

    if request.guest_count == 0 {
        return Err(BookingRejected::InvalidGuestCount);
    }

    if request.stay.check_in < now.date_naive() {
        return Err(BookingRejected::StayStartsInPast);
    }

    match context.guest {
        None => return Err(BookingRejected::GuestNotFound),
        Some(GuestBookingEligibility::Blocked) => return Err(BookingRejected::GuestBlocked),
        Some(GuestBookingEligibility::Eligible) => {}
    }

    let listing = match &context.listing {
        None => return Err(BookingRejected::ListingNotFound),
        Some(listing) => listing,
    };

    validate_listing(request, listing)?;

    if context
        .unavailable_nights
        .iter()
        .any(|night| *night >= request.stay.check_in && *night < request.stay.check_out)
    {
        return Err(BookingRejected::ListingUnavailable);
    }

    Ok(ReservationConfirmed {
        reservation_id,
        guest_id: request.guest_id,
        listing_id: request.listing_id,
        stay: request.stay,
        guest_count: request.guest_count,
        confirmed_at: now,
    })
}

fn validate_listing(
    request: &BookStay,
    listing: &ListingBookingSettings,
) -> Result<(), BookingRejected> {
    if listing.booking_status == ListingBookingStatus::Disabled {
        return Err(BookingRejected::ListingDisabled);
    }

    if request.guest_count > listing.max_guests {
        return Err(BookingRejected::TooManyGuests);
    }

    let stay_nights = request
        .stay
        .check_out
        .signed_duration_since(request.stay.check_in)
        .num_days();
    let min_nights = i64::from(listing.min_nights);
    let max_nights = listing.max_nights.map(i64::from);

    if stay_nights < min_nights {
        return Err(BookingRejected::StayTooShort);
    }

    if max_nights.is_some_and(|max_nights| stay_nights > max_nights) {
        return Err(BookingRejected::StayTooLong);
    }

    Ok(())
}

File: csharp-example/src/BookingAStay/Capabilities/BookStay/Decide.cs
namespace BookingAStay.Capabilities.BookStay;

public static class Decider
{
    public static BookingOutcome Decide(
        BookStayRequest request,
        BookingContext context,
        Guid reservationId,
        DateTimeOffset now)
    {
        if (request.Stay.CheckIn >= request.Stay.CheckOut)
        {
            return new BookingOutcome.Rejected(BookingRejection.InvalidDateRange);
        }

        if (request.GuestCount <= 0)
        {
            return new BookingOutcome.Rejected(BookingRejection.InvalidGuestCount);
        }

        if (request.Stay.CheckIn < DateOnly.FromDateTime(now.UtcDateTime))
        {
            return new BookingOutcome.Rejected(BookingRejection.StayStartsInPast);
        }

        switch (context.Guest)
        {
            case null:
                return new BookingOutcome.Rejected(BookingRejection.GuestNotFound);
            case GuestBookingEligibility.Blocked:
                return new BookingOutcome.Rejected(BookingRejection.GuestBlocked);
            case GuestBookingEligibility.Eligible:
                break;
        }

        var listing = context.Listing;
        if (listing is null)
        {
            return new BookingOutcome.Rejected(BookingRejection.ListingNotFound);
        }

        var listingRejection = ValidateListing(request, listing);
        if (listingRejection is not null)
        {
            return new BookingOutcome.Rejected(listingRejection.Value);
        }

        if (context.UnavailableNights.Any(night => night >= request.Stay.CheckIn && night < request.Stay.CheckOut))
        {
            return new BookingOutcome.Rejected(BookingRejection.ListingUnavailable);
        }

        return new BookingOutcome.Confirmed(new ReservationConfirmed(
            reservationId,
            request.GuestId,
            request.ListingId,
            request.Stay,
            request.GuestCount,
            now,
            listing.MaxGuests,
            listing.MinNights,
            listing.MaxNights));
    }

    private static BookingRejection? ValidateListing(BookStayRequest request, ListingBookingSettings listing)
    {
        if (listing.BookingStatus == ListingBookingStatus.Disabled)
        {
            return BookingRejection.ListingDisabled;
        }

        if (request.GuestCount > listing.MaxGuests)
        {
            return BookingRejection.TooManyGuests;
        }

        var stayNights = request.Stay.CheckOut.DayNumber - request.Stay.CheckIn.DayNumber;

        if (stayNights < listing.MinNights)
        {
            return BookingRejection.StayTooShort;
        }

        if (listing.MaxNights is { } maxNights && stayNights > maxNights)
        {
            return BookingRejection.StayTooLong;
        }

        return null;
    }
}
